src/write_mongodb.c

```diff
@@ -55,179 +51,239 @@ struct wm_node_s {
   char *passwd;
 
   _Bool store_rates;
+  bool connected;
 
-  mongo conn[1];
+  mongoc_client_t *client;
+  mongoc_database_t *database;
   pthread_mutex_t lock;
 };
 typedef struct wm_node_s wm_node_t;
 
 /*
  * Functions
  */
-static bson *wm_create_bson(const data_set_t *ds, /* {{{ */
-                            const value_list_t *vl, _Bool store_rates) {
-  bson *ret;
+static bson_t *wm_create_bson(const data_set_t *ds, /* {{{ */
+                              const value_list_t *vl, _Bool store_rates) {
+  bson_t *ret;
+  bson_t subarray;
   gauge_t *rates;
 
-  ret = bson_alloc(); /* matched by bson_dealloc() */
-  if (ret == NULL) {
-    ERROR("write_mongodb plugin: bson_create failed.");
-    return (NULL);
+  ret = bson_new();
+  if (!ret) {
+    ERROR("write_mongodb plugin: bson_new failed.");
+    return NULL;
   }
 
   if (store_rates) {
     rates = uc_get_rate(ds, vl);
     if (rates == NULL) {
       ERROR("write_mongodb plugin: uc_get_rate() failed.");
-      return (NULL);
+      bson_free(ret);
+      return NULL;
     }
   } else {
     rates = NULL;
   }
 
-  bson_init(ret); /* matched by bson_destroy() */
-  bson_append_date(ret, "time", (bson_date_t)CDTIME_T_TO_MS(vl->time));
-  bson_append_string(ret, "host", vl->host);
-  bson_append_string(ret, "plugin", vl->plugin);
-  bson_append_string(ret, "plugin_instance", vl->plugin_instance);
-  bson_append_string(ret, "type", vl->type);
-  bson_append_string(ret, "type_instance", vl->type_instance);
+  BSON_APPEND_DATE_TIME(ret, "timestamp", CDTIME_T_TO_MS(vl->time));
+  BSON_APPEND_UTF8(ret, "host", vl->host);
```

Comment:
I added a validation check at the end of the method that makes sure all strings are valid UTF-8 strings.